18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,115 @@
+import React, { useCallback, useEffect, useState, useRef } from 'react';
+import io from 'socket.io-client';
+
+import Message from './components/Message';
+
+import './Chat.css';
+
+const socketUrl = 'http://localhost:3002';
+const client = io(socketUrl);
+
+function Chat() {
+  const socket = useRef(client).current;
+  const [status, setStatus] = useState('Disconnected');
+  const [inputValue, setInputValue] = useState('');
+  const [username, setUsername] = useState('anon');
+  const [messages, setMessages] = useState([]);
+  const [typingUsers, setTypingUsers] = useState(new Set());
+
+  const addMessage = useCallback((msg) => {
+    setMessages([...messages, msg]);
+  }, [messages, setMessages]);
+
+  const addTyping = useCallback((data) => {
+    setTypingUsers((typing) => new Set([...typing, data.username]));
+  }, [setTypingUsers]);
+
+  const removeTyping = useCallback((data) => {
+    setTypingUsers((typing) => {
+      typing.delete(data.username);
+      return typing;
+    });
+  }, []);
+
+
+  useEffect(() => {
+    if (socket) {
+      socket.on('message', (msg) => {
+        switch (msg.type) {
+          case 'message':
+            addMessage(msg);
+            removeTyping(msg);
+            break;
+          case 'typing':
+            addTyping(msg);
+            break;
+          default:
+            break;
+        }
+      });
+
+      setStatus(socket.status);
+    }
+  }, [addMessage, setStatus, addTyping, socket, removeTyping]);
+
+  const sendMessage = useCallback((event) => {
+    event.preventDefault();
+    setInputValue('');
+    socket.emit('message', { type: 'message', payload: inputValue });
+  }, [inputValue, setInputValue, socket]);
+
+  const changeUsername = useCallback((event) => {
+    event.preventDefault();
+    socket.emit('message', { type: 'set_username', payload: username });
+  }, [username, socket]);
+
+  const handleInputChange = useCallback((event) => {
+    setInputValue(event.target.value);
+    socket.emit('message', { type: 'typing' });
```

Comment:
if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the `debounce` method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.

Reply:
fixed with lodash's debounce